File: python/backtest_double_EMA_StochRSI_float.py
import pandas as pd
import numpy as np
import pandas_ta as ta
import psutil
import time
import os

abspath = os.path.abspath(__file__)
dname = os.path.dirname(abspath)
os.chdir(dname)

class RUN_RESULT():
    pass

DATAFILE = "../data/Binance/4h/BTC-USDT.csv"
STRAT_NAME = "2-EMA crossover with Stoch RSI"

start_year = 2017  # forced year to start (applies if data below is available)
FEE = 0.07        # FEES in %
USDT_amount_initial = 1000.0
MIN_NUMBER_OF_TRADES_PER_YEAR = 14 # minimum number of trades required (to avoid some noise / lucky circunstances)
MIN_NUMBER_OF_TRADES = -1
MIN_ALLOWED_MAX_DRAWBACK = -50.0  # %
MAX_ALLOWED_DAYS_BETWEEN_PORTFOLIO_ATH = 365
STOCH_RSI_UPPER = 0.800
STOCH_RSI_LOWER = 0.200
minimum_yearly_gain_pc = -100.0  # pc
i_start_year = 0

# RANGE OF EMA PERIDOS TO TEST ####################################
period_max_EMA = 350
range_step = 1
range1 = np.arange(2, period_max_EMA, range_step)
range2 = np.arange(2, period_max_EMA, range_step)
#range1 = [116]
#range2 = [3]
###################################################################

i_print = 0
ii_begin = 0
nb_tested = 0

###################################################################

def custom_stochRSI_TravingView_Style(close, length=14, rsi_length=14, k=3, d=3):
    """Indicator: Stochastic RSI Oscillator (STOCHRSI)
    Should be similar to TradingView's calculation"""
    if k<0:
        raise Exception("k cannot be negative")
    if d<0:
        raise Exception("d cannot be negative")
    # Calculate Result
    rsi_ = ta.rsi(close, length=rsi_length,talib=False)
    lowest_rsi = rsi_.rolling(length).min()
    highest_rsi = rsi_.rolling(length).max()
    stochrsi = 100.0 * (rsi_ - lowest_rsi) / ta.non_zero_range(highest_rsi, lowest_rsi)
    if k>0:
        stochrsi_k = ta.ma('sma', stochrsi, length=k ,talib=False)
        stochrsi_d = ta.ma('sma', stochrsi_k, length=d ,talib=False)
    else:
        stochrsi_k = None
        stochrsi_d = None
    return (stochrsi/100.0).round(4), (stochrsi_k/100.0).round(4), (stochrsi_d/100.0).round(4)

###################################################################

def print_res(best):
    print("-------------------------------------")
    print("EMA: ", best.ema1, " ", best.ema2)
    print("Gain: ", best.gain_pc, "%")
    print("Win rate: ", best.win_rate, "%")
    print("max DD: ", best.max_DD, "%")
    print("Gain over DDC: ", best.gain_over_DDC)
    print("Score: ", best.score)
    print("Number of trades: ", best.nb_posi_entered)
###################################################################

def print_best_res(best):
    print("\n-------------------------------------")
    print("BEST PARAMETER SET FOUND: ")
    print("-------------------------------------")
    print(f"Strategy : {STRAT_NAME}")
    print(f"EMAs     : ", best.ema1, " ", best.ema2)
    print(f"Best Gain: {best.gain_pc:.2f}%")
    print(f"Porfolio : {best.WALLET_VAL_USDT:.2f}$ (started with 1000$)")
    print(f"Win rate : {best.win_rate:.2f}%")
    print(f"max DD   : {best.max_DD:.2f}%")
    print(f"Gain/DDC : {best.gain_over_DDC:.2f}")
    print(f"Score    : {best.score:.2f}")
    print("Number of trades: ", best.nb_posi_entered)
    print("Yearly gains: ")
    for i in range(len(best.years_yearly_gains)):
        print(best.years_yearly_gains[i], " :: ", best.yearly_gains[i], "%")

    print("Total fees paid: ", round(best.total_fees_paid*100.0)/100.0, "$ (started with 1000$)")

    print("-------------------------------------")

###################################################################

def INITIALIZE_DATA(kline):
    global MIN_NUMBER_OF_TRADES
    global ii_begin
    list_ema = []

    maxval = np.max([np.max(range1), np.max(range2)]) + 5
    for i in range(maxval):
        list_ema.append(i)

    for i in list_ema:
        kline["EMA" + str(i)] = ta.ema(kline["close"], length=int(i), talib=False)

    print("Calculated EMAs.")
    kline['StochRSI'], _, _ = custom_stochRSI_TravingView_Style(kline["close"], length=14, rsi_length=14, k=3, d=3)
    print("Calculated STOCHRSI.")

    kline['date'] = pd.to_datetime(kline['time'], unit='ms')

    kline['year']  = kline['date'].dt.year
    kline['hour']  = kline['date'].dt.hour
    kline['month'] = kline['date'].dt.month
    kline['day']   = kline['date'].dt.day
    kline['shifted_year']  = kline['year'].shift(-1)

    MIN_NUMBER_OF_TRADES = MIN_NUMBER_OF_TRADES_PER_YEAR * int(np.max(kline['year']) - np.min(kline['year']) + 1)

    ii_begin = period_max_EMA + 2

    print("Initialized calculations.")

###################################################################

def PROCESS(kline, ema1_v, ema2_v):
    global nb_tested
    global range1
    global range2
    global i_print
    global ii_begin
    global FEE

    nb_tested += 1

    result = RUN_RESULT()

    result.years_yearly_gains = []
    result.yearly_gains = []

    nb_max = len(kline)

    LAST_ITERATION = False
    OPEN_LONG_CONDI = False
    CLOSE_LONG_CONDI = False
    nb_profit = 0
    nb_loss = 0
    NB_POSI_ENTERED = 0
    pc_change_with_max = 0.0
    max_drawdown = 0.0
    price_position_open = 0.0

    USDT_amount = USDT_amount_initial
    WALLET_VAL_begin_year = USDT_amount_initial
    MAX_WALLET_VAL_USDT = USDT_amount_initial
    COIN_AMOUNT = 0.0
    total_fees_paid_USDT = 0.0
    WALLET_VAL_USDT = USDT_amount_initial

    emastr1 = "EMA" + str(ema1_v)
    emastr2 = "EMA" + str(ema2_v)

    for ii, row in kline.iterrows():

        if (ii<ii_begin):
            continue

        if (ii == nb_max - 1):
            LAST_ITERATION = True

        # condition for open / close position
        OPEN_LONG_CONDI  = row[emastr2] >= row[emastr1] and row['StochRSI'] > STOCH_RSI_UPPER
        CLOSE_LONG_CONDI = row[emastr2] <= row[emastr1] and row['StochRSI'] < STOCH_RSI_LOWER

        # IT IS IMPORTANT TO CHECK FIRST FOR CLOSING POSITION AND THEN FOR OPENING POSITION

        # CLOSE LONG
        if (COIN_AMOUNT > 0.0 and (CLOSE_LONG_CONDI or LAST_ITERATION)):

            USDT_amount = COIN_AMOUNT * row['close']
            COIN_AMOUNT = 0.0

            # apply FEEs
            fe = USDT_amount * FEE / 100.0
            USDT_amount -= fe
            total_fees_paid_USDT += fe
            #
            if (row['close'] >= price_position_open):
                nb_profit += 1
            else:
                nb_loss += 1

        # OPEN LONG
        if (COIN_AMOUNT == 0.0 and OPEN_LONG_CONDI and LAST_ITERATION == False):

            price_position_open = row['close']
            COIN_AMOUNT = USDT_amount / row['close']
            USDT_amount = 0.0

            # apply FEEs
            fe = COIN_AMOUNT * FEE / 100.0
            COIN_AMOUNT -= fe
            total_fees_paid_USDT += fe * row['close']
            #

            NB_POSI_ENTERED += 1

        # check yealy gains

        if (row['shifted_year'] != row['year'] or LAST_ITERATION):
            result.years_yearly_gains.append(row['year'])
            WALLET_VAL_USDT = USDT_amount + COIN_AMOUNT * row['close']
            yg = (WALLET_VAL_USDT - WALLET_VAL_begin_year) / WALLET_VAL_begin_year * 100.0
            result.yearly_gains.append(np.round(yg * 100.0) / 100.0)
            WALLET_VAL_begin_year = WALLET_VAL_USDT

        # check wallet status
        if (CLOSE_LONG_CONDI or LAST_ITERATION):
            WALLET_VAL_USDT = USDT_amount + COIN_AMOUNT * row['close']
            if (WALLET_VAL_USDT > MAX_WALLET_VAL_USDT):
                MAX_WALLET_VAL_USDT = WALLET_VAL_USDT

            pc_change_with_max = (WALLET_VAL_USDT - MAX_WALLET_VAL_USDT) / MAX_WALLET_VAL_USDT * 100.0

            if (pc_change_with_max < max_drawdown):
                max_drawdown = pc_change_with_max

    WALLET_VAL_USDT = USDT_amount + COIN_AMOUNT * row['close']

    gain = (WALLET_VAL_USDT - USDT_amount_initial) / USDT_amount_initial * 100.0
    WR = float(nb_profit) / float(NB_POSI_ENTERED) * 100.0
    DDC = (1.0 / (1.0 + max_drawdown / 100.0) - 1.0) * 100.0
    score = gain / DDC * WR

    i_print += 1

    if (i_print == 1000):
        i_print = 0
        print("DONE: EMA: ", ema1_v, " and EMA: ", ema2_v)

    result.WALLET_VAL_USDT = USDT_amount
    result.gain_over_DDC = gain / DDC
    result.gain_pc = gain
    result.max_DD = max_drawdown
    result.nb_posi_entered = NB_POSI_ENTERED
    result.win_rate = WR
    result.score = score
    result.ema1 = ema1_v
    result.ema2 = ema2_v
    result.total_fees_paid = total_fees_paid_USDT

    return result

###################################################################

def read_input_data(input_file_path):
    kline = pd.read_csv(input_file_path, delimiter=';',header=None,names=['time','open','high','low','close','volume'])
    print(kline.tail(4))
    print("Loaded data file.")
    return kline

###################################################################

if __name__ == "__main__":

    t_begin = time.time()

    print("\n-------------------------------------")
    print("Strategy to test: ", STRAT_NAME)
    print("DATA FILE TO PROCESS: ", DATAFILE)

    kline = read_input_data(DATAFILE)

    INITIALIZE_DATA(kline)

    best = RUN_RESULT()

    best.gain_over_DDC = -100.0

    # Display info
    print("Begin day      : ", kline['year'][0], "/", kline['month'][0], "/", kline['day'][0])
    print("End day        : ", kline['year'].iloc[-1], "/", kline['month'].iloc[-1], "/", kline['day'].iloc[-1])
    print("OPEN/CLOSE FEE : ", FEE, " %")
    print("Minimum number of trades required    : ", MIN_NUMBER_OF_TRADES)
    print("Maximum drawback (=drawdown) allowed : ", MIN_ALLOWED_MAX_DRAWBACK, " %")
    print("StochRSI Upper Band   : ", STOCH_RSI_UPPER)
    print("StochRSI Lower Band   : ", STOCH_RSI_LOWER)
    print("EMA period max tested : ", period_max_EMA)
    print("EMA range step        : ", range_step)
    print("-------------------------------------")

    # MAIN LOOP

    for ema1 in range1:
        for ema2 in range2:

            if (np.abs(ema1-ema2) < 3):
                continue

            res = PROCESS(kline, ema1, ema2)

            if (res.gain_over_DDC > best.gain_over_DDC
                    and res.gain_pc < 100000.0
                    and res.nb_posi_entered >= MIN_NUMBER_OF_TRADES
                    and res.max_DD > MIN_ALLOWED_MAX_DRAWBACK
                    and np.min(res.yearly_gains) > minimum_yearly_gain_pc):

                best = res

    print_best_res(best)
    t_end = time.time()

    print("Number of backtests performed : ", nb_tested)
    print("Time taken                    : ", int(t_end - t_begin), " seconds ")
    ram_usage = psutil.virtual_memory().available * 100.0 / psutil.virtual_memory().total
    print("RAM usage                     : ", np.round(ram_usage * 10.0) / 10.0 , " MB" )
    print("-------------------------------------")

File: backtest_double_EMA_StochRSI_float.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <fstream>
#include <sstream>
#include <math.h>
#include <unordered_map>
#include "tools.hh"
#include "custom_talib_wrapper.hh"
#include <ta-lib/ta_libc.h>
using namespace std;

const std::string DATAFILE = "./data/Binance/30m/BTC-USDT.csv";
const std::string STRAT_NAME = "2-EMA crossover with Stoch RSI";

const float start_year = 2017; // forced year to start (applies if data below is available)
const float FEE = 0.05;        // FEES in %
const float USDT_amount_initial = 1000.0;
const int MIN_NUMBER_OF_TRADES_PER_YEAR = 14; // minimum number of trades required (to avoid some noise / lucky circunstances)
int MIN_NUMBER_OF_TRADES = -1;                // minimum number of trades required (to avoid some noise / lucky circunstances)
const float MIN_ALLOWED_MAX_DRAWBACK = -50.0; // %
const float MAX_ALLOWED_DAYS_BETWEEN_PORTFOLIO_ATH = 365;
const float STOCH_RSI_UPPER = 0.800;
const float STOCH_RSI_LOWER = 0.200;
const float minimum_yearly_gain_pc = -100.0; // pc
int i_start_year = 0;

// RANGE OF EMA PERIDOS TO TESTs
const int period_max_EMA = 350;
const int range_step = 1;
std::vector<int> range1 = integer_range(2, period_max_EMA, range_step);
std::vector<int> range2 = integer_range(2, period_max_EMA, range_step);
//////////////////////////

uint i_print = 0;

std::unordered_map<std::string, std::vector<float>> EMA_LISTS{};
std::vector<float> StochRSI{};
std::vector<float> year{};
std::vector<float> hour{};
std::vector<float> month{};
std::vector<float> day{};
uint nb_tested = 0;

//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

void print_res(const RUN_RESULTf best)
{
    std::cout << "-------------------------------------" << std::endl;
    std::cout << "EMA: " << best.ema1 << " " << best.ema2 << std::endl;
    std::cout << "Gain: " << best.gain_pc << "%" << std::endl;
    std::cout << "Win rate: " << best.win_rate << "%" << std::endl;
    std::cout << "max DD: " << best.max_DD << "%" << std::endl;
    std::cout << "Gain over DDC: " << best.gain_over_DDC << std::endl;
    std::cout << "Score: " << best.score << std::endl;
    std::cout << "Number of trades: " << best.nb_posi_entered << std::endl;
}
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

void print_best_res(const RUN_RESULTf best)
{
    std::cout << "\n-------------------------------------" << std::endl;
    std::cout << "BEST PARAMETER SET FOUND: " << std::endl;
    std::cout << "-------------------------------------" << std::endl;
    std::cout << "Strategy : " << STRAT_NAME << std::endl;
    std::cout << "EMAs     : " << best.ema1 << " " << best.ema2 << std::endl;
    std::cout << "Best Gain: " << best.gain_pc << "%" << std::endl;
    std::cout << "Porfolio : " << best.WALLET_VAL_USDT << "$ (started with 1000$)" << std::endl;
    std::cout << "Win rate : " << best.win_rate << "%" << std::endl;
    std::cout << "max DD   : " << best.max_DD << "%" << std::endl;
    std::cout << "Gain/DDC : " << best.gain_over_DDC << std::endl;
    std::cout << "Score    : " << best.score << std::endl;
    std::cout << "Number of trades: " << best.nb_posi_entered << std::endl;
    std::cout << "Max days between porfolio ATH: " << int(float(best.max_delta_t_new_ATH) / 3600.0 / 24.0) << std::endl;
    std::cout << "Yearly gains: " << std::endl;
    for (uint i = 0; i < best.yearly_gains.size(); i++)
    {
        std::cout << best.years_yearly_gains[i] << " :: " << best.yearly_gains[i] << "%" << std::endl;
    }
    std::cout << "Total fees paid: " << round(best.total_fees_paid*100.0)/100.0 << "$ (started with 1000$)" << std::endl;

    std::cout << "-------------------------------------" << std::endl;
}

//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
void INITIALIZE_DATA(const KLINEf &kline)
{
    std::vector<int> list_ema = {};

    for (uint i = 2; i <= std::max(find_max(range1), find_max(range2)) + 5; i++)
    {
        list_ema.push_back(i);
    }

    for (const uint i : list_ema)
    {
        EMA_LISTS["EMA" + std::to_string(i)] = TALIB_EMA(kline.d_close, i);
    }
    cout << "Calculated EMAs." << endl;
    //StochRSI = TALIB_STOCHRSI_K(kline.d_close,14,3,3);
    StochRSI = TALIB_STOCHRSI_not_averaged(kline.d_close, 14, 14);
    cout << "Calculated STOCHRSI." << endl;

    float yy_b = 1990;
    for (uint ii = 0; ii < kline.nb; ii++)
    {
        const float yy = get_year_from_timestamp(kline.d_time[ii]);
        year.push_back(yy);
        hour.push_back(get_hour_from_timestamp(kline.d_time[ii]));
        month.push_back(get_month_from_timestamp(kline.d_time[ii]));
        day.push_back(get_day_from_timestamp(kline.d_time[ii]));
        // if (ii<100) cout << StochRSI[ii] << endl;
        if (yy >= start_year && yy_b < start_year)
        {
            i_start_year = ii;
        }
        yy_b = yy;
    }

    MIN_NUMBER_OF_TRADES = MIN_NUMBER_OF_TRADES_PER_YEAR * int(find_max(year) - find_min(year) + 1);

    cout << "Initialized calculations." << endl;
}

//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

RUN_RESULTf PROCESS(const KLINEf &KLINEf, const int ema1_v, const int ema2_v)
{
    nb_tested++;

    RUN_RESULTf result{};

    std::vector<float> close = KLINEf.d_close;
    std::vector<int> timestamp = KLINEf.d_time;
    std::vector<float> EMA1 = EMA_LISTS["EMA" + std::to_string(ema1_v)];
    std::vector<float> EMA2 = EMA_LISTS["EMA" + std::to_string(ema2_v)];

    const int nb_max = KLINEf.nb;

    bool LAST_ITERATION = false;
    bool OPEN_LONG_CONDI = false, CLOSE_LONG_CONDI = false;
    int nb_profit = 0, nb_loss = 0, NB_POSI_ENTERED = 0;
    float pc_change_with_max = 0, max_drawdown = 0, price_position_open = 0;
    float USDT_amount = USDT_amount_initial;
    float WALLET_VAL_begin_year = USDT_amount_initial;
    float MAX_WALLET_VAL_USDT = USDT_amount_initial;
    float COIN_AMOUNT = 0.0;
    float total_fees_paid_USDT = 0.0;
    float WALLET_VAL_USDT = USDT_amount_initial;

    int t_new_ATH = 0;
    int delta_t_new_ATH = 0;
    int max_delta_t_new_ATH = 0;

    int ii_begin = period_max_EMA + 2;
    ii_begin = std::max(i_start_year, ii_begin);

    for (uint ii = ii_begin; ii < nb_max; ii++)
    {
        if (ii == nb_max - 1) LAST_ITERATION = true;

        // condition for open / close position
        OPEN_LONG_CONDI = EMA2[ii] >= EMA1[ii] && StochRSI[ii] > STOCH_RSI_UPPER;
        CLOSE_LONG_CONDI = EMA2[ii] <= EMA1[ii] && StochRSI[ii] < STOCH_RSI_LOWER;

        // IT IS IMPORTANT TO CHECK FIRST FOR CLOSING POSITION AND THEN FOR OPENING POSITION

        // CLOSE LONG
        if (COIN_AMOUNT>0.0 && (CLOSE_LONG_CONDI || LAST_ITERATION))
        {
            USDT_amount = COIN_AMOUNT * close[ii];
            COIN_AMOUNT = 0.0;

            // apply FEEs
            const float fe = USDT_amount * FEE / 100.0;
            USDT_amount -= fe;
            total_fees_paid_USDT += fe;
            //
            if (close[ii] >= price_position_open)
            {
                nb_profit++;
            }
            else
            {
                nb_loss++;
            }
        }

        // OPEN LONG
        if (COIN_AMOUNT==0.0 && OPEN_LONG_CONDI && LAST_ITERATION==false)
        {
            price_position_open = close[ii];

            COIN_AMOUNT = USDT_amount / close[ii];
            USDT_amount = 0.0;

            // apply FEEs
            const float fe = COIN_AMOUNT * FEE / 100.0;
            COIN_AMOUNT -= fe;
            total_fees_paid_USDT += fe * close[ii];
            //

            NB_POSI_ENTERED++;
        }

        // check yealy gains
        if (year[ii - 1] != year[ii] || ii == KLINEf.nb - 1)
        {
            result.years_yearly_gains.push_back(year[ii - 1]);
            WALLET_VAL_USDT = USDT_amount + COIN_AMOUNT * close[ii];
            const float yg = (WALLET_VAL_USDT - WALLET_VAL_begin_year) / WALLET_VAL_begin_year * 100.0;
            result.yearly_gains.push_back(std::round(yg * 100.0) / 100.0);
            WALLET_VAL_begin_year = WALLET_VAL_USDT;
        }

        // check wallet status
        if (CLOSE_LONG_CONDI || LAST_ITERATION) {
            WALLET_VAL_USDT = USDT_amount + COIN_AMOUNT * close[ii];
            if (WALLET_VAL_USDT > MAX_WALLET_VAL_USDT) MAX_WALLET_VAL_USDT = WALLET_VAL_USDT;
            pc_change_with_max = (WALLET_VAL_USDT - MAX_WALLET_VAL_USDT) / MAX_WALLET_VAL_USDT * 100.0;
            if (pc_change_with_max < max_drawdown) max_drawdown = pc_change_with_max;
        
            if (t_new_ATH != 0)
                {
                    delta_t_new_ATH = timestamp[ii] - t_new_ATH;
                    if (delta_t_new_ATH > max_delta_t_new_ATH)
                    {
                        max_delta_t_new_ATH = delta_t_new_ATH;
                    }
                }
                t_new_ATH = timestamp[ii];
        }
    }

    WALLET_VAL_USDT = USDT_amount + COIN_AMOUNT * close[close.size() - 1];

    const float gain = (WALLET_VAL_USDT - USDT_amount_initial) / USDT_amount_initial * 100.0;
    const float WR = float(nb_profit) / float(NB_POSI_ENTERED) * 100.0;
    const float DDC = (1.0 / (1.0 + max_drawdown / 100.0) - 1.0) * 100.0;
    const float score = gain / DDC * WR;

    i_print++;
    if (i_print == 30000)
    {
        i_print = 0;
        std::cout << "DONE: EMA: " << ema1_v << " and EMA: " << ema2_v << endl;
    }

    result.WALLET_VAL_USDT = USDT_amount;
    result.gain_over_DDC = gain / DDC;
    result.gain_pc = gain;
    result.max_DD = max_drawdown;
    result.nb_posi_entered = NB_POSI_ENTERED;
    result.win_rate = WR;
    result.score = score;
    result.ema1 = ema1_v;
    result.ema2 = ema2_v;
    result.total_fees_paid = total_fees_paid_USDT;
    result.max_delta_t_new_ATH = max_delta_t_new_ATH;

    return result;
}

//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

KLINEf read_input_data(const std::string &input_file_path)
{
    KLINEf kline;

    ifstream myfile(input_file_path);
    std::string line;
    long int ts;
    float op, hi, lo, cl, vol;
    int ho;
    char comma = ';';
    bool first_line = true;

    ifstream file(input_file_path);
    string value;
    vector<string> getInform;
    while (file.good())
    {
        getline(file, value);
        const auto pos = value.find(';');
        value = ReplaceAll(value, ";", " ");
        getInform.push_back(value);
        // getInform.erase(getInform.begin()+1);
        stringstream ss{};
        ss << value;
        ss >> ts >> op >> hi >> lo >> cl >> vol;
        kline.d_time.push_back(ts / 1000);
        kline.d_open.push_back(op);
        kline.d_high.push_back(hi);
        kline.d_low.push_back(lo);
        kline.d_close.push_back(cl);
        // cout << kline.d_open.back()<<endl;
    }

    myfile.close();

    kline.nb = int(kline.d_close.size());

    std::cout << "Loaded data file." << std::endl;
    return kline;
}

//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

int main()
{
    // std::cout << range_step << std::endl;
    // std::cout << period_max_EMA << std::endl; 
    // std::vector<float> test{1.,2.,3.,4.,5.,6.,7.,8.,9.,10.,11.,12.,13.,14.,15.,16.,17.,5.,4.,3.,2.,1.};
    // std::vector<float> EMA = TALIB_EMA(test,14);
    // std::vector<float> RSI = TALIB_RSI(test,14);
    // std::vector<float> MIN = TALIB_MIN(test,14);
    // std::cout << MIN.size() << std::endl;
    // for (uint i = 0; i < RSI.size(); i++)
    // {
    //     cout << RSI[i] << endl;
    // }
    // std::abort();   

    double t_begin = get_wall_time();
    std::cout << "\n-------------------------------------" << std::endl;
    std::cout << "Strategy to test: " << STRAT_NAME << std::endl;
    std::cout << "DATA FILE TO PROCESS: " << DATAFILE << std::endl;

    TA_RetCode retCode;
    retCode = TA_Initialize();
    if (retCode != TA_SUCCESS)
    {
        std::cout << "Cannot initialize TA-Lib !\n"
                  << retCode << "\n";
    }
    else
    {
        std::cout << "Initialized TA-Lib !\n";
    }

    const KLINEf kline = read_input_data(DATAFILE);

    INITIALIZE_DATA(kline);

    RUN_RESULTf best{};
    best.gain_over_DDC = -100.0;

    int i_print2 = 0;

    // Display info
    std::cout << "Begin day      : " << year[i_start_year] << "/" << month[i_start_year] << "/" << day[i_start_year] << std::endl;
    std::cout << "End day        : " << year.back() << "/" << month.back() << "/" << day.back() << std::endl;
    std::cout << "OPEN/CLOSE FEE : " << FEE << " %" << std::endl;
    std::cout << "Minimum number of trades required    : " << MIN_NUMBER_OF_TRADES << std::endl;
    std::cout << "Maximum drawback (=drawdown) allowed : " << MIN_ALLOWED_MAX_DRAWBACK << " %" << std::endl;
    std::cout << "StochRSI Upper Band   : " << STOCH_RSI_UPPER << std::endl;
    std::cout << "StochRSI Lower Band   : " << STOCH_RSI_LOWER << std::endl;
    std::cout << "EMA period max tested : " << period_max_EMA << std::endl;
    std::cout << "EMA range step        : " << range_step << std::endl;
    std::cout << "-------------------------------------" << std::endl;

    // MAIN LOOP

    for (int ema1 : range1)
    {
        for (int ema2 : range2)
        {
            if (std::abs(ema1-ema2)<3) continue;

            RUN_RESULTf res = PROCESS(kline, ema1, ema2);

            if (res.gain_over_DDC > best.gain_over_DDC 
                && res.gain_pc < 100000.0 
                && res.nb_posi_entered >= MIN_NUMBER_OF_TRADES // should do at least 100 trades
                && res.max_DD > MIN_ALLOWED_MAX_DRAWBACK 
                && find_min(res.yearly_gains) > minimum_yearly_gain_pc)
            {
                best = res;
            }
        }
    }

    print_best_res(best);
    double t_end = get_wall_time();

    std::cout << "Number of backtests performed : " << nb_tested << std::endl;
    std::cout << "Time taken                    : " << t_end - t_begin << " seconds " << std::endl;
    const double ram_usage = process_mem_usage();
    std::cout << "RAM usage                     : " << std::round(ram_usage * 10.0) / 10.0 << " MB" << std::endl;
    std::cout << "-------------------------------------" << std::endl;

    TA_Shutdown();
}
